cpp/src/merge/merge.cu

```diff
@@ -342,17 +339,65 @@ std::unique_ptr<column> column_merger::operator()<cudf::dictionary32>(
 {
   auto result = cudf::dictionary::detail::merge(
     cudf::dictionary_column_view(lcol), cudf::dictionary_column_view(rcol), row_order_, stream, mr);
+
   // set the validity mask
   if (lcol.has_nulls() || rcol.has_nulls()) {
     auto merged_view = result->mutable_view();
-    materialize_bitmask(lcol, rcol, merged_view, row_order_.data(), stream);
+    materialize_bitmask(
+      lcol, rcol, merged_view.null_mask(), merged_view.size(), row_order_.data(), stream);
   }
   return result;
 }
 
+// specialization for structs
+template <>
+std::unique_ptr<column> column_merger::operator()<cudf::struct_view>(
+  column_view const& lcol,
+  column_view const& rcol,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr) const
+{
+  // merge each child.
+  std::vector<std::unique_ptr<column>> merged_children;
+  merged_children.reserve(lcol.num_children());
+
+  structs_column_view lhs(lcol);
+  structs_column_view rhs(rcol);
+
+  column_merger merger{row_order_};
+  auto iter = thrust::make_counting_iterator(0);
+  std::transform(
+    iter, iter + lhs.num_children(), std::back_inserter(merged_children), [&](size_type i) {
+      return cudf::type_dispatcher<dispatch_storage_type>(
+        lhs.child(i).type(), merger, lhs.get_sliced_child(i), rhs.get_sliced_child(i), stream, mr);
+    });
+
+  auto const merged_size = lcol.size() + rcol.size();
```

Comment:
You can use a 1cudf::detail::make_counting_transform_iterator` to clean this up a little bit.
```suggestion
  auto const lhs = structs_column_view{lcol};
  auto const rhs = structs_column_view{rcol};

  auto it = cudf::detail::make_counting_transform_iterator(
    0, [&, merger = column_merger{row_order_}](size_type i) {
      return cudf::type_dispatcher<dispatch_storage_type>(
        lhs.child(i).type(), merger, lhs.get_sliced_child(i), rhs.get_sliced_child(i), stream, mr);
    });

  auto merged_children   = std::vector<std::unique_ptr<column>>(it, it + lhs.num_children());
  auto const merged_size = lcol.size() + rcol.size();
```

Reply:
I don't want this cleanup to hold up the PR so I will isolate it.